pl_bolts/models/self_supervised/simsiam/simsiam_module.py

```diff
@@ -0,0 +1,484 @@
+        self.batch_size = batch_size
+
+        self.hidden_mlp = hidden_mlp
+        self.feat_dim = feat_dim
+        self.first_conv = first_conv
+        self.maxpool1 = maxpool1
+
+        self.optim = optimizer
+        self.lars_wrapper = lars_wrapper
+        self.exclude_bn_bias = exclude_bn_bias
+        self.weight_decay = weight_decay
+        self.temperature = temperature
+
+        self.start_lr = start_lr
+        self.final_lr = final_lr
+        self.learning_rate = learning_rate
+        self.warmup_epochs = warmup_epochs
+        self.max_epochs = max_epochs
+
+        self.init_model()
+
+        # compute iters per epoch
+        global_batch_size = self.nodes * self.gpus * self.batch_size if self.gpus > 0 else self.batch_size
+        self.train_iters_per_epoch = self.num_samples // global_batch_size
+
+        # define LR schedule
+        warmup_lr_schedule = np.linspace(
+            self.start_lr, self.learning_rate, self.train_iters_per_epoch * self.warmup_epochs
+        )
+        iters = np.arange(self.train_iters_per_epoch * (self.max_epochs - self.warmup_epochs))
+        cosine_lr_schedule = np.array([self.final_lr + 0.5 * (self.learning_rate - self.final_lr) * (
+            1 + math.cos(math.pi * t / (self.train_iters_per_epoch * (self.max_epochs - self.warmup_epochs)))
+        ) for t in iters])
+
+        self.lr_schedule = np.concatenate((warmup_lr_schedule, cosine_lr_schedule))
+
+    def init_model(self):
+        if self.arch == 'resnet18':
+            backbone = resnet18
+        elif self.arch == 'resnet50':
+            backbone = resnet50
+
+        encoder = backbone(
+            first_conv=self.first_conv, maxpool1=self.maxpool1, return_all_feature_maps=False
+        )
+        self.online_network = SiameseArm(encoder, input_dim=self.hidden_mlp,
+                                         hidden_size=self.hidden_mlp, output_dim=self.feat_dim)
+        self.init_target_network()
+
+    def init_target_network(self):
+        self.target_network = deepcopy(self.online_network)
+
+    def on_train_batch_end(self, outputs, batch: Any, batch_idx: int, dataloader_idx: int) -> None:
+        self.init_target_network()
+
+    def forward(self, x):
+        y, _, _ = self.online_network(x)
+        return y
+
+    def cosine_similarity(self, a, b):
+        b = b.detach()  # stop gradient of backbone + projection mlp
+        a = F.normalize(a, dim=-1)
+        b = F.normalize(b, dim=-1)
+        sim = (a * b).sum(-1).mean()
+        return sim
+
+    def shared_step(self, batch, batch_idx):
+        (img_1, img_2, _), y = batch
+
+        # Image 1 to image 2 loss
+        _, z1, h1 = self.online_network(img_1)
+        _, z2, h2 = self.target_network(img_2)
+        loss_a = -1.0 * self.cosine_similarity(h1, z2)
+
+        # Image 2 to image 1 loss
+        _, z1, h1 = self.online_network(img_2)
+        _, z2, h2 = self.target_network(img_1)
+        loss_b = -1.0 * self.cosine_similarity(h1, z2)
```

Comment:
Based on SimSiam's pseudocode and other [implementations](https://github.com/PatrickHua/SimSiam/blob/main/models/simsiam.py), shouldn't there be just one network without a deep copy target network?

Reply:
Believe it is equivalent. For example if you refer to BYOL implementation in pl bolts there is a deep copy here there isn't https://github.com/lucidrains/byol-pytorch

Reply:
Nonetheless, great job on the implementation. It might be worth investigating in the future if there is any difference in performance between the two methods. As with this approach, maybe there is more memory usage?

Reply:
With a deep copy instead of using the same network twice you use ~3GB gpu extra memory with resnet18. The two versions are about equally fast on cifar10, but I think that's because the gpu is memorybound on the task.
Given that the final performane ends up the same, I guess removing the copied network would be a good idea as it's less wasteful and scales better.

Reply:
Thanks @MikkelAntonsen.
Could you please share the version you ran with plot of results on CIFAR10?

Reply:
https://gist.github.com/MikkelAntonsen/d390d128b28b06a7fc6ed9d2e73b6378
In short, functions init_target_network, on_train_batch_end, shared_step and attribute target_network have been removed. Training_step has been updated accordingly.

![bilde](https://user-images.githubusercontent.com/2040712/103171356-eb28d180-484b-11eb-98e7-ab7ad07a1523.png)

Reply:
Thanks for sharing @MikkelAntonsen.
I just pushed a commit with the changes you suggested in the gist.
Great job making the improvements in efficiency while maintaining accuracy.

Reply:
I noticed that the SSLOnlineEvaluator does an additional forward pass per train_batch to get embeddings. It would be possible to return the encoder output from `training_step()` and which would be accessible in the `outputs` argument in SSLOnlineEvaluator, AFAIK. This is unfortunate because it couples the implementation of SSLOnlineEvaluator with the networks that uses it. But if we are to reproduce results on imagenet, is reducing the number of forward passes by 1/3 negligible?

Reply:
Hi @MikkelAntonsen I believe this is related more to SSL capabilities in general of pl_bolts so might be better to open a separate issue since this isn't only effecting SimSiam (tagging @ananyahjha93).

Reply:
@MikkelAntonsen Maybe the byol implementation can also benefit from using only one network (without deepcopy) and using detach() to control the gradient flow?

Reply:
I just read the BYOL paper abstract and it seems like the online network and target network use a different set of weights, so I'm not sure how you could share a network. If you look at figure 2 in the paper, it does seem like they use the stop gradient trick, but only for the target network. Do you see any ways to incorporate simsiam ideas into BYOL without the implemention just ending up as simsiam?